pkg/controller/eventinstance/truncate.go

```diff
@@ -0,0 +1,163 @@
+			// This ensures it's started after the router cache has synced
+			go func() {
+				if err := t.start(ctx); err != nil {
+					logrus.WithError(err).Warn("Exiting eventinstance truncator")
+				}
+			}()
+		})
+
+		eventKey := router.Key(req.Namespace, req.Name)
+		if req.Object != nil {
+			// Add or Update, add to the running count of events
+			t.track(eventKey)
+		} else {
+			// Delete, remove from the running count of events
+			t.forget(eventKey)
+		}
+
+		return nil
+	}
+}
+
+type truncator struct {
+	client   kclient.Client
+	capacity int
+	tracking map[kclient.ObjectKey]struct{}
+	trigger  chan struct{}
+	mu       *sync.Mutex
+}
+
+func (t *truncator) start(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.trigger:
+		}
+
+		if err := t.truncate(ctx); err != nil {
+			logrus.WithError(err).Warn("Failed to truncate eventinstances, retrying in 15 seconds")
+			go func() {
+				time.Sleep(15 * time.Second)
+				t.kick()
+			}()
+			continue
+		}
+
+		// Give the router a chance to process more events
+		time.Sleep(1 * time.Minute)
+	}
+}
+
+func (t *truncator) kick() {
+	select {
+	case t.trigger <- struct{}{}:
+	default:
+	}
+}
+
+func (t *truncator) track(eventKey kclient.ObjectKey) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.tracking[eventKey] = struct{}{}
+
+	if len(t.tracking) > t.capacity {
+		t.kick()
+	}
+}
+
+func (t *truncator) forget(eventKey kclient.ObjectKey) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	delete(t.tracking, eventKey)
+}
+
+func (t *truncator) truncate(ctx context.Context) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if len(t.tracking) <= t.capacity {
+		// Nothing to do
+		return nil
+	}
+
+	var events internalv1.EventInstanceList
+	if err := t.client.List(ctx, &events); err != nil {
+		return err
+	}
+
+	if len(events.Items) <= t.capacity {
+		// The client says we're under capacity, so reset tracking to match
+		t.tracking = make(map[kclient.ObjectKey]struct{}, len(events.Items))
+		for _, e := range events.Items {
+			t.tracking[kclient.ObjectKeyFromObject(&e)] = struct{}{}
+		}
+
+		return nil
+	}
+
+	// Sort events by observed so that the newest events are at the beginning of the list
+	sort.Slice(events.Items, func(i, j int) bool {
+		return events.Items[i].Observed.After(events.Items[j].Observed.Time)
+	})
+
+	t.tracking = make(map[kclient.ObjectKey]struct{}, t.capacity)
+	var errs []error
+	for i, e := range events.Items {
+		if i < t.capacity {
+			t.tracking[kclient.ObjectKeyFromObject(&e)] = struct{}{}
+			continue
+		}
+
+		if err := hardDelete(ctx, t.client, &e); err != nil && !apierrors.IsNotFound(err) {
+			t.tracking[kclient.ObjectKeyFromObject(&e)] = struct{}{}
+			errs = append(errs, err)
+			continue
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+func hardDelete(ctx context.Context, c kclient.Client, e *internalv1.EventInstance) error {
+	if len(e.Finalizers) > 0 {
+		// Clear all finalizers before deletion
+		e.Finalizers = nil
```

Comment:
There will be when I reuse this controller in the account-controller in manager. I think for this particular controller it makes sense to use a big hammer -- if we didn't a rogue finalizer can cause events to grow unbounded

Reply:
The finalizers on these objects in manager ensure that the deletion is seen in the downstream cluster. Additionally, updating the object will will cause the controller in the cluster-agent to re-add the finalizer and may lead to the delete here failing.

Reply:
> The finalizers on these objects in manager ensure that the deletion is seen in the downstream cluster.

Ah, interesting point.

Which also has me thinking that maybe this would all be simpler if this logic only runs in upstream?

Reply:
It probably would be simpler if it ran upstream only, but I definitely think it can run in both places. Even if it only ran in `runtime`, then I think it would mostly achieve the goal.

Reply:
OK -- both it is

I dropped the finalizer update btw.